File: packages/client/src/memmachine_client/format.py
"""LLM-friendly formatting for memory search and list results.

These functions mirror the server's internal formatting logic so that
client-side consumers get the same compact representation the server
uses when feeding memories into language models.

Episodic format matches ``string_from_episode_context`` in the server.
Semantic format matches ``_features_to_llm_format`` in the server.
"""

from __future__ import annotations

import json
from collections.abc import Iterable

from memmachine_common.api.spec import (
    Episode,
    EpisodeResponse,
    SearchResult,
    SemanticFeature,
)


def format_episodes(episodes: Iterable[EpisodeResponse | Episode]) -> str:
    """Format episodic memories as an LLM-friendly string.

    Each episode is rendered as::

        [Monday, January 01, 2024 at 01:30 PM] producer_id: "content"

    This mirrors the server's ``string_from_episode_context`` output.

    Args:
        episodes: Episodic memory entries from a search or list result.

    Returns:
        Newline-terminated formatted string (empty string when *episodes*
        is empty).

    """
    result = ""
    for episode in episodes:
        if episode.created_at is not None:
            date_str = episode.created_at.strftime("%A, %B %d, %Y")
            time_str = episode.created_at.strftime("%I:%M %p")
            result += f"[{date_str} at {time_str}] {episode.producer_id}: {json.dumps(episode.content)}\n"
        else:
            result += f"{episode.producer_id}: {json.dumps(episode.content)}\n"
    return result


def format_semantic_memories(features: Iterable[SemanticFeature]) -> str:
    """Format semantic memories as a compact JSON string.

    Produces a ``{tag: {feature_name: value}}`` structure, omitting all
    metadata for context efficiency.  This mirrors the server's
    ``_features_to_llm_format`` output.

    Args:
        features: Semantic memory entries from a search or list result.

    Returns:
        JSON string of the grouped features.

    """
    structured: dict[str, dict[str, str]] = {}
    for feature in features:
        structured.setdefault(feature.tag, {})[feature.feature_name] = feature.value
    return json.dumps(structured)


def format_search_result(result: SearchResult) -> str:
    """Format a search result as an LLM-friendly string.

    Combines episodic and semantic memories from a
    :class:`~memmachine_common.api.spec.SearchResult` returned by
    ``Memory.search()``.

    Args:
        result: A ``SearchResult`` object.

    Returns:
        Formatted string combining both memory types.

    """
    sections: list[str] = []

    if result.content.episodic_memory is not None:
        episodes = (
            result.content.episodic_memory.long_term_memory.episodes
            + result.content.episodic_memory.short_term_memory.episodes
        )
        if episodes:
            sections.append(f"[Episodic Memory]\n{format_episodes(episodes)}")

    if result.content.semantic_memory:
        sections.append(
            f"[Semantic Memory]\n{format_semantic_memories(result.content.semantic_memory)}"
        )

    return "\n".join(sections)

File: packages/ts-client/src/memory/format.ts
/**
 * LLM-friendly formatting for memory search and list results.
 *
 * These functions mirror the server's internal formatting logic so that
 * client-side consumers get the same compact representation the server
 * uses when feeding memories into language models.
 *
 * Episodic format matches `string_from_episode_context` on the server.
 * Semantic format matches `_features_to_llm_format` on the server.
 *
 * @packageDocumentation
 */

import type { SemanticMemory, SearchMemoriesResult } from './memmachine-memory.types'

/**
 * Minimal shape required to format an episode.
 *
 * Both search-response episodes ({@link EpisodicMemory}) and list-response
 * episodes ({@link ListEpisodicMemory}) satisfy this shape, so
 * {@link formatEpisodes} works with episodes from either endpoint.
 */
export interface FormattableEpisode {
  created_at?: string | null
  producer_id: string
  content: string
}

const DAYS = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'] as const
const MONTHS = [
  'January',
  'February',
  'March',
  'April',
  'May',
  'June',
  'July',
  'August',
  'September',
  'October',
  'November',
  'December'
] as const

function _formatDate(date: Date): string {
  const day = DAYS[date.getUTCDay()]
  const month = MONTHS[date.getUTCMonth()]
  const dd = String(date.getUTCDate()).padStart(2, '0')
  const year = date.getUTCFullYear()
  return `${day}, ${month} ${dd}, ${year}`
}

function _formatTime(date: Date): string {
  let hours = date.getUTCHours()
  const minutes = String(date.getUTCMinutes()).padStart(2, '0')
  const ampm = hours >= 12 ? 'PM' : 'AM'
  hours = hours % 12 || 12
  return `${String(hours).padStart(2, '0')}:${minutes} ${ampm}`
}

/**
 * Format episodic memories as an LLM-friendly string.
 *
 * Each episode is rendered as:
 * ```
 * [Monday, January 01, 2024 at 01:30 PM] producer_id: "content"
 * ```
 *
 * This mirrors the server's `string_from_episode_context` output.
 *
 * Accepts any object with `created_at`, `producer_id`, and `content`,
 * so both search-response and list-response episodes work without
 * additional adaptation.
 *
 * @param episodes - Episodic memory entries from a search or list result.
 * @returns Newline-terminated formatted string (empty string when episodes is empty).
 */
export function formatEpisodes(episodes: FormattableEpisode[]): string {
  let result = ''
  for (const episode of episodes) {
    if (episode.created_at) {
      const date = new Date(episode.created_at)
      const dateStr = _formatDate(date)
      const timeStr = _formatTime(date)
      result += `[${dateStr} at ${timeStr}] ${episode.producer_id}: ${JSON.stringify(episode.content)}\n`
    } else {
      result += `${episode.producer_id}: ${JSON.stringify(episode.content)}\n`
    }
  }
  return result
}

/**
 * Format semantic memories as a compact JSON string.
 *
 * Produces a `{tag: {feature_name: value}}` structure, omitting all
 * metadata for context efficiency. This mirrors the server's
 * `_features_to_llm_format` output.
 *
 * @param features - Semantic memory entries from a search or list result.
 * @returns JSON string of the grouped features.
 */
export function formatSemanticMemories(features: SemanticMemory[]): string {
  const structured = Object.create(null) as Record<string, Record<string, string>>
  for (const feature of features) {
    if (!(feature.tag in structured)) {
      structured[feature.tag] = Object.create(null) as Record<string, string>
    }
    structured[feature.tag]![feature.feature_name] = feature.value
  }
  return JSON.stringify(structured)
}

/**
 * Format a search result as an LLM-friendly string.
 *
 * Combines episodic and semantic memories from a {@link SearchMemoriesResult}
 * returned by `MemMachineMemory.search()`.
 *
 * @param result - A `SearchMemoriesResult` object.
 * @returns Formatted string combining both memory types.
 */
export function formatSearchResult(result: SearchMemoriesResult): string {
  const sections: string[] = []

  if (result.content.episodic_memory) {
    const episodes = [
      ...result.content.episodic_memory.long_term_memory.episodes,
      ...result.content.episodic_memory.short_term_memory.episodes
    ]
    if (episodes.length > 0) {
      sections.push(`[Episodic Memory]\n${formatEpisodes(episodes)}`)
    }
  }

  if (result.content.semantic_memory && result.content.semantic_memory.length > 0) {
    sections.push(`[Semantic Memory]\n${formatSemanticMemories(result.content.semantic_memory)}`)
  }

  return sections.join('\n')
}
